### PR TITLE
feat: add fill functions for revision history

### PR DESCRIPTION
**Pull request by MarcoKorinth**
# Fillfunctions in /document/tracking

## Changes
1. Add fill function for `Date of revision` (Clock icon) in Tracking/RevisionHistory/*/
2. Add fill function for `current_release_date` in Tracking/Fields
3. Add fill function for `initial_release_date` in Tracking/Fields

## Steps to test
- In FormEditor navigate to Tracking
- Add new Revision history Item (3 dots, "Add list item")
  - Set Version to 1.0.0
  - Set date e.g. 01.01.2020
- Add second Revision history Item
  - Set Version later than 1 e.g. 2.0.0
  - Set current date with clock icon (should enter current date and time rounded to next hour)
- Navigate to Tracking/Fields
  - Set initial release and current release with fill functions (magic wand icon)
  - Both should be filled with the previously set values

### CODE REVIEW

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 1]
```diff
@@ -607,4 +607,202 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+      ).click()
+      cy.get(
+        '[data-testid="attribute-document-tracking-initial_release_date"] input[type="date"]'
+      ).should('have.value', '2020-01-01')
+      cy.get(
+        '[data-testid="attribute-document-tracking-initial_release_date"] input[type="time"]'
+      ).should('have.value', '12:00')
+    })
+  })
+})
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:808]: The case with more than one Revision History entry is missing.
=> Initial release must not change, Current release must take the new time

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 2]
```diff
@@ -607,4 +607,202 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+      ).click()
+      cy.get(
+        '[data-testid="attribute-document-tracking-revision_history-0-date"] input[type="date"]'
+      )
+        .invoke('attr', 'value', '2020-01-01')
+        .trigger('change')
+      cy.get(
+        '[data-testid="attribute-document-tracking-revision_history-0-date"] input[type="time"]'
+      )
+        .invoke('attr', 'value', '12:00')
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:768]: Using a "full-hour" value here seems to be counter-productive as it wouldn't detect if a rounding function was used somewhere => use 09:41 instead.

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Please see my feedback.

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 3]
```diff
@@ -607,4 +607,202 @@ describe('SecvisogramPage / FormEditor Tab', function () {
       '[data-testid="attribute-product_tree-branches-0-branches-0-product-name"] input'
     ).should('have.value', 'Vendor A Product DEF')
   })
-})
+
+  describe('fill functions for revision history', function () {
+    it('fill date of revision', function () {
+      cy.visit('?tab=EDITOR')
+      const now = new Date(2020, 0, 1, 10, 30)
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:614]: I'm still wondering - is the `0` correct in the second place?

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 4]
```diff
@@ -607,4 +607,202 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+      ).click()
+      cy.get(
+        '[data-testid="attribute-document-tracking-revision_history-0-date"] input[type="date"]'
+      ).should('have.value', '2020-01-01')
+      cy.get(
+        '[data-testid="attribute-document-tracking-revision_history-0-date"] input[type="time"]'
+      ).should('have.value', '11:00')
+    })
+
+    it('fill current release date', function () {
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:657]: This test fails for me locally.

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 5]
```diff
@@ -607,4 +607,202 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+      cy.get(
+        '[data-testid="attribute-document-tracking-current_release_date"] input[type="date"]'
+      ).should('have.value', '2020-01-01')
+      cy.get(
+        '[data-testid="attribute-document-tracking-current_release_date"] input[type="time"]'
+      ).should('have.value', '12:00')
+    })
+
+    it('fill initial release date', function () {
+      cy.visit('?tab=EDITOR')
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:731]: This test fails for me locally.

File: app/lib/app/SecvisogramPage/View/FormEditor/shared/fillFieldFunctions.js [hunk 6]
```diff
@@ -86,4 +86,52 @@ const getRelationshipName = async function (
...
-export { uniqueProductId, uniqueGroupId, getBranchName, getRelationshipName }
+/**
+ * function to get the current Date rounded to the next full hour
+ *
+ * @return string|undefined
+ */
+const getCurrentDateRounded = function () {
+  const p = 60 * 60 * 1000 // milliseconds in an hour
+  const roundedDate = new Date(Math.ceil(new Date().getTime() / p) * p)
+  return roundedDate.toISOString()
```
[the remark this answers is not in the document]
Comment (MarcoKorinth) [reply app/lib/app/SecvisogramPage/View/FormEditor/shared/fillFieldFunctions.js:97]: IMO `toISOString` returns a valid RFC3339 string.
Or am I missing something?
Example: `2023-10-24T11:48:44.063Z`

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 5 above)
Comment (MarcoKorinth) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:731]: both tests work for me locally and on github.
Could you comment your error message?

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 3 above)
Comment (MarcoKorinth) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:614]: Yes, month indexes start at 0 for the js Date constructor.
See: [Date() constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#monthindex)

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 2 above)
Comment (MarcoKorinth) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:768]: Yes I agree. I'll change that

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 5 above)
Comment (MarcoKorinth) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:729]: I changed the way date and time values are invoked, maybe that fixed the issue for you

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 4 above)
Comment (tschmidtb51) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:657]: Works now.

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js (hunk 5 above)
Comment (tschmidtb51) [reply app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:729]: Works now. Thanks.

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 7]
```diff
@@ -607,4 +607,268 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+      cy.get(
+        '[data-testid="menu_entry-/document/tracking/revision_history-add_item_button"]'
+      ).click({ force: true })
+      cy.get(
+        '[data-testid="menu_entry-/document/tracking/revision_history/1"]'
+      ).click()
+      cy.get(
+        '[data-testid="attribute-document-tracking-revision_history-1-date"] input[type="date"]'
+      )
+        .type('2021-01-01')
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:839]: As a method of caution: Please use a date here that also differs in month and date, e.g. 2021-02-28

File: app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js [hunk 8]
```diff
@@ -607,4 +607,268 @@ describe('SecvisogramPage / FormEditor Tab', function () {
...
+        '[data-testid="attribute-document-tracking-initial_release_date"] input[type="date"]'
+      ).should('have.value', '2020-01-01')
+      cy.get(
+        '[data-testid="attribute-document-tracking-initial_release_date"] input[type="time"]'
+      ).should('have.value', '13:41')
+    })
+
+    it('fill release dates with more than one revision history entry', function () {
+      cy.visit('?tab=EDITOR')
+
```
Comment (tschmidtb51) [inline app/cypress/e2e/SecvisogramPage/FormEditorTab.cy.js:806]: Please also add a version `0.9.0` (2019-12-31T12:01:00Z) here to check that the right date is selected

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Some minor things - other than that ok.

**Review by tschmidtb51 — APPROVED** [review]
LGTM

**Review by StefanNemeth — APPROVED** [review]

**Pull request merged**